builtin/providers/openstack/resource_openstack_images_image_v2.go

```diff
@@ -0,0 +1,488 @@
+	hash := md5.New()
+	if _, err := io.Copy(hash, f); err != nil {
+		return "", err
+	}
+	return hex.EncodeToString(hash.Sum(nil)), nil
+}
+
+func resourceImagesImageV2FileProps(filename string) (int64, string, error) {
+	var filesize int64
+	var filechecksum string
+
+	file, err := os.Open(filename)
+	if err != nil {
+		return -1, "", fmt.Errorf("Error opening file for Image: %s", err)
+
+	}
+	defer file.Close()
+
+	fstat, err := file.Stat()
+	if err != nil {
+		return -1, "", fmt.Errorf("Error reading image file %q: %s", file.Name(), err)
+	}
+
+	filesize = fstat.Size()
+	filechecksum, err = fileMD5Checksum(file)
+
+	if err != nil {
+		return -1, "", fmt.Errorf("Error computing image file %q checksum: %s", file.Name(), err)
+	}
+
+	return filesize, filechecksum, nil
+}
+
+func resourceImagesImageV2File(d *schema.ResourceData) (string, error) {
+	if filename := d.Get("local_file_path").(string); filename != "" {
+		return filename, nil
+	} else if furl := d.Get("image_source_url").(string); furl != "" {
+		dir := d.Get("image_cache_path").(string)
+		os.MkdirAll(dir, 0700)
+		filename := filepath.Join(dir, fmt.Sprintf("%x.img", md5.Sum([]byte(furl))))
+
+		if _, err := os.Stat(filename); err != nil {
+			if !os.IsNotExist(err) {
+				return "", fmt.Errorf("Error while trying to access file %q: %s", filename, err)
+			}
+			log.Printf("[DEBUG] File doens't exists %s. will download from %s", filename, furl)
+			file, err := os.Create(filename)
+			if err != nil {
+				return "", fmt.Errorf("Error creating file %q: %s", filename, err)
+			}
+			defer file.Close()
+			resp, err := http.Get(furl)
+			if err != nil {
+				return "", fmt.Errorf("Error downloading image from %q", furl)
+			}
+			defer resp.Body.Close()
+
+			if _, err = io.Copy(file, resp.Body); err != nil {
+				return "", fmt.Errorf("Error downloading image %q to file %q: %s", furl, filename, err)
+			}
+			return filename, nil
+		} else {
+			log.Printf("[DEBUG] File exists %s", filename)
+			return filename, nil
+		}
+	} else {
+		return "", fmt.Errorf("Error in config. no file specified")
+	}
+}
+
+func resourceImagesImageV2Read(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	imageClient, err := config.imageV2Client(d.Get("region").(string))
+	if err != nil {
+		return fmt.Errorf("Error creating OpenStack image client: %s", err)
+	}
+
+	img, err := images.Get(imageClient, d.Id()).Extract()
+	if err != nil {
+		return CheckDeleted(d, err, "image")
+	}
+
+	log.Printf("[DEBUG] Retrieved Image %s: %#v", d.Id(), img)
+
+	d.Set("owner", img.Owner)
+	d.Set("status", img.Status)
+	d.Set("file", img.File)
+	d.Set("schema", img.Schema)
+	d.Set("checksum", img.Checksum)
+	d.Set("size_bytes", img.SizeBytes)
+	d.Set("metadata", img.Metadata)
+	d.Set("created_at", img.CreatedAt)
+	d.Set("update_at", img.UpdatedAt)
+	d.Set("id", img.ID)
+
+	return nil
+}
+
+func resourceImagesImageV2Update(d *schema.ResourceData, meta interface{}) error {
```

Comment:
This function is correctly placed below `resourceImagesImageV2Read`, but make sure to move this up when moving `Read`.